Lib/fontTools/varLib/mutator.py

```diff
@@ -30,6 +30,69 @@
 	OS2_WIDTH_CLASS_VALUES[half] = i
 
 
+def interpolate_cff2_PrivateDict(topDict, applyScalars):
+	pd_blend_lists = ("BlueValues", "OtherBlues", "FamilyBlues",
+						"FamilyOtherBlues", "StemSnapH",
+						"StemSnapV")
+	pd_blend_values = ("BlueScale", "BlueShift",
+						"BlueFuzz", "StdHW", "StdVW")
+	for fontDict in topDict.FDArray:
+		pd = fontDict.Private
+		vsindex = pd.vsindex if (hasattr(pd, 'vsindex')) else 0
+		for key, value in pd.rawDict.items():
+			if (key in pd_blend_values) and isinstance(value, list):
+					deltas = value[1:]
+					pd.rawDict[key] = otRound(value[0] + applyScalars(vsindex, deltas))
+			elif (key in pd_blend_lists) and isinstance(value[0], list):
+				"""If any argument in a BlueValues list is a blend list,
+				then they all are. The first value of each list is an
+				absolute value. The delta tuples are calculated from
+				relative master values, hence we need to append all the
+				deltas to date to each successive absolute value."""
+				delta = 0
+				for i, val_list in enumerate(value):
+					delta += otRound(applyScalars(vsindex, val_list[1:]))
+					value[i] = val_list[0] + delta
+
+
+def interpolate_cff2_charstrings(topDict, applyScalars, glyphOrder):
+	charstrings = topDict.CharStrings
+	for gname in glyphOrder:
+		charstring = charstrings[gname]
+		charstring.decompile()
+		vsindex = charstring.private.vsindex if (
+													hasattr(charstring.private,
+													'vsindex')) else 0
+		num_regions = charstring.private.getNumRegions(vsindex)
+		numMasters = num_regions + 1
+		new_program = []
+		last_i = 0
+		for i, token in enumerate(charstring.program):
```

Comment:
So, this assumes no subroutines?

Reply:
Whoops. I spent last week with T2CharstringPen.draw(), where the flattening is all taken care of, and forgot I was workng with non-flattened charstrings. I certainly need to allow for subroutines.

Reply:
Done. Let me know if you would like the CFF2 support functions moved from mutator.py to elsewhere, or anther approach to flattening the charstrings - there are lots of ways to do this.

Reply:
> Done.

I'm confused.  Don't we need to flatten recursively?  Isn't this exactly the `_DesubroutinizingT2Decompiler` in fontTools.subset.__init__.py?  Let's move that somewhere else and use it.  I also prefer that naming to "flattening" which can be ambiguous.

Also, won't we need drop to drop subroutines from the font?  Ie. exactly the desubroutinization operation.

> Let me know if you would like the CFF2 support functions moved from mutator.py to elsewhere, or anther approach to flattening the charstrings - there are lots of ways to do this.

Those can stay for now.

Reply:
This code needs to be moved to a module under cffLib and reused:
https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/subset/__init__.py#L2204-L2600

@anthrotype ideas for naming?

Reply:
fontToola.cffLib.decompilers

Reply:
Humm.  While true that these are decompilers, I guess the API to them will be just functions, like `font['CFF '].cff.desubroutinize()` or `.drop_hints()`, since there's some work to do after the decompiler.  The internals can be exposed (with no API guarantee) for those who need more access.  Oh, one other useful function, that is used after both operations, is to remove unused subroutines.  This involves renumbering all other subroutines and call sites.  Is a very useful optimization.  The subsetter calls those functionalities `prune`.

Same about code that removes / renumbers unused lookups in GSUB/GPOS.  We should move that code somewhere reusable.  For example, varLib.mutator needs to call that after applying `FeatureVariations` such that unused lookups are dropped.

Reply:
Maybe it's fine to expose these from submodules under the subset module?

Reply:
> Maybe it's fine to expose these from submodules under the subset module?

sure

Reply:
> Maybe it's fine to expose these from submodules under the subset module?

Subset?  I'm not ready to move interpolation-space-subsetting (of which mutator is a special-case of) to subset yet.  I suggest exposing them from varLib.mutator.* instead.

Reply:
well, if they are needed both by subset and varLib, then cffLib would be the best candidate.
anyway, i'm ok either way.

Reply:
Oh.  I confused myself again.  The github email showed your comment in reply to the code, which is very varLib-specific.  Now I see you were replying to my proposal.  Yeah, I think submodules in subset are best right now.  I thought about cffLib, but then some more will go in GSUB/GPOS, etc.  That's one way to order thing.  The other we have been using is to group by functionality.  So let's stick to the latter for now.

Reply:
I was not aware of _DesubroutinizingT2Decompiler in fontTools.subset. Certainly no need to re-invent the wheel. I would like to tweak _DesubroutinizingT2Decompiler some: CFF2 charstrings have neither endchar not return ops, so lines 2431-2436 are not needed for that charstring flavor. I would also like to at least have a mode where the charString is not desubroutinized every time it is called. Some CJK fonts max out at the limit of 64K subrs , most of which get called many times, so this would add up to serious overhead.

Reply:
I think the naming discussion about concluded that _DesubroutinizingT2Decompiler and everything else on L2204-2600 should stay in in fontTools.subset, and that interpolate_cff2_PrivateDict and interpolate_cff2_charstrings should stay in mutator.py, is that correct? 

Also, interpolate_cff2_charstrings does drop Subrs from the font, rather crudely by simply deleting them all, which is fine because we flatten all the charstrings.

Reply:
> I would also like to at least have a mode where the charString is not desubroutinized every time it is called.

Right.  When I wrote the code originally I did it this way, to ensure correctness.  Code seems to have performed solidly in the years since, so I'm fine using the cache now.

Reply:
> Also, interpolate_cff2_charstrings does drop Subrs from the font, rather crudely by simply deleting them all, which is fine because we flatten all the charstrings.

Correct.  But then again, the desubroutinize() function should do that (the code in subsetter does).

So, let's introduce `subset.cff` and `varLib.cff` modules?

Reply:
@behdad I switched to using a new desubroutinize() function that I added to subset.__init__.py - take a look.

After looking at subset.__init.py, I'd prefer to extract the CFF-related functions from subset.__init__.py into a new subset.cffLib module after we finish and merge the changes needed for this PR. I see a lot of lines which are CFF specific ( vs CFF2) that I'd like to fix at the same time, and I think it would be easier to review as a separate branch and PR.

Reply:
Sure.  `subset.cff` please, no Lib suffix.